pkg/apis/kf/v1alpha1/app_types.go

```diff
@@ -0,0 +1,158 @@
+type App struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +optional
+	Spec AppSpec `json:"spec,omitempty"`
+
+	// +optional
+	Status AppStatus `json:"status,omitempty"`
+}
+
+// AppSpec is the desired configuration for an App.
+type AppSpec struct {
+
+	// Source contains the source configuration of the App.
+	// +optional
+	Source AppSpecSource `json:"source,omitempty"`
+
+	// Template defines the App's desired runtime configuration.
+	// +optional
+	Template AppSpecTemplate `json:"template,omitempty"`
+
+	// Routes defines network routes for the App's ingress.
+	// +optional
+	Routes AppSpecRoutes `json:"routes,omitempty"`
+
+	// Services defines what services the App requires.
+	// +optional
+	Services AppSpecServices `json:"services,omitempty"`
+
+	// Serve determines if the App should be running.
+	Serve *bool `json:"serve"`
+}
+
+// AppSpecSource defines the source code for an App.
+// The fields ContainerImage and BuildpackBuild are mutually exclusive.
+type AppSpecSource struct {
+
+	// UId is a unique identifier for an AppSpecSource.
+	// Updating sub-values will trigger a new value.
+	// +optional
+	UId string `json:"uid,omitempty"`
+
+	// ContainerImage defines the container image for source.
+	// +optional
+	ContainerImage AppSpecSourceContainerImage `json:"containerimage,omitempty"`
+
+	// BuildpackBuild defines buildpack information for source.
+	// +optional
+	BuildpackBuild AppSpecSourceBuildpackBuild `json:"buildpackbuild,omitempty"`
+}
+
+// AppSpecSourceContainerImage defines a container image for an App.
+type AppSpecSourceContainerImage struct {
+
+	// Image is the container image URI for the App.
+	Image string `json:"image"`
+}
+
+// AppSpecSourceBuildpackBuild defines building an App using Buildpacks.
+type AppSpecSourceBuildpackBuild struct {
+
+	// Source is the Container Image which contains the App's source code.
+	// +optional
+	Source string `json:"source,omitempty"`
+
+	// Stack is the base layer to use for the App.
+	// +optional
+	Stack string `json:"stack,omitempty"`
+
+	// Buildpack is the Buildpack to use for the App.
+	// +optional
+	Buildpack string `json:"buildpack,omitempty"`
+}
+
+// AppSpecTemplate defines an App's desired runtime configuration.
+type AppSpecTemplate struct {
+
+	// UId is a unique identifier for an AppSpecTemplate.
+	// Updating sub-values will trigger a new value.
+	// +optional
+	UId string `json:"uid,omitempty"`
+
+	// Scale defines the App's scaling configuration.
+	// +optional
+	Scale AppSpecTemplateScale `json:"scale,omitempty"`
+
+	// Spec is a PodSpec with additional restrictions.
+	// The image name is ignored.
+	// The Spec contains configuration for the App's Pod.
+	// (Env, Vars, Quotas, etc)
+	// +optional
+	Spec core.PodSpec `json:"spec,omitempty"`
+}
+
+// AppSpecTemplateScale defines an App's scaling configuration.
+type AppSpecTemplateScale struct {
```

Comment:
@poy had some comments about this in the issue, WDYT about consolidating this with the status to produce a unioned field like the following and stuck at the spec level? 

```
instances:
  stopped: # true if the service and routes should stop traffic
  exactly: # exactly this number can run
  min: # minimum of this number
  max: # maximum of this number, 0 = infinite
```

That would be used like this:

```
instances: # scale to n
  exactly: 3

instances: # scale to 0
  max: 5

instances: # stopped
  stopped: true
  min: 1
  max: 5
```

Reply:
Should be addressed